File: sdk/go/client/evaluate_contract_test.go
package client

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestEvaluateDecisionWithScopeBindsHeadersAndCanonicalBody(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/api/v1/evaluate" {
			t.Fatalf("request = %s %s", r.Method, r.URL.Path)
		}
		for header, want := range map[string]string{
			"Authorization":       "Bearer token",
			"X-Helm-Tenant-ID":    "tenant-a",
			"X-Helm-Principal-ID": "operator-a",
			"X-Helm-Session-ID":   "session-a",
			"X-Helm-Workspace-ID": "workspace-a",
			"Idempotency-Key":     "evaluate-1",
		} {
			if got := r.Header.Get(header); got != want {
				t.Fatalf("%s = %q, want %q", header, got, want)
			}
		}
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body["action"] != "read" || body["resource"] != "ticket:1" {
			t.Fatalf("body = %#v", body)
		}
		if _, exists := body["principal"]; exists {
			t.Fatalf("caller identity leaked into request body: %#v", body)
		}
		history, ok := body["session_history"].([]interface{})
		if !ok || len(history) != 1 {
			t.Fatalf("session history = %#v", body["session_history"])
		}
		entry, ok := history[0].(map[string]interface{})
		if !ok || entry["action"] != "read-history" || entry["resource"] != "ticket:0" || entry["verdict"] != "ALLOW" || entry["timestamp"] != float64(1) {
			t.Fatalf("session history entry = %#v", history[0])
		}
		if _, exists := entry["principal"]; exists {
			t.Fatalf("caller identity leaked into session history: %#v", entry)
		}
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("X-Helm-Receipt-ID", "rcpt-decision-1")
		w.Header().Set("X-Helm-Idempotency-Replayed", "true")
		_, _ = w.Write([]byte(`{"id":"decision-1","tenant_id":"tenant-a","session_id":"session-a","signature_schema":"helm.decision.signature.v2"}`))
	}))
	defer server.Close()

	client := New(
		server.URL,
		WithAPIKey("token"),
		WithTenantID("tenant-default"),
		WithPrincipalID("principal-default"),
		WithWorkspaceID("workspace-default"),
	)
	request := DecisionRequest{
		Action:   "read",
		Resource: "ticket:1",
		SessionHistory: []SessionAction{{
			Action:               "read-history",
			Resource:             "ticket:0",
			Verdict:              "ALLOW",
			Timestamp:            1,
			AdditionalProperties: map[string]interface{}{"principal": "must-not-serialize"},
		}},
	}
	request.AdditionalProperties = map[string]interface{}{"principal": "must-not-serialize"}
	result, err := client.EvaluateDecisionWithScope(request, EvaluationScope{
		TenantID:    "tenant-a",
		PrincipalID: "operator-a",
		SessionID:   "session-a",
		WorkspaceID: "workspace-a",
	}, "evaluate-1")
	if err != nil {
		t.Fatal(err)
	}
	if result.Decision.GetId() != "decision-1" || result.ReceiptID != "rcpt-decision-1" || !result.Replayed {
		t.Fatalf("result = %#v", result)
	}
}

func TestEvaluateDecisionWithScopeFailsLocallyWithoutRequiredBinding(t *testing.T) {
	client := New("http://127.0.0.1:1", WithAPIKey("token"))
	_, err := client.EvaluateDecisionWithScope(DecisionRequest{Action: "read", Resource: "ticket:1"}, EvaluationScope{
		TenantID:  "tenant-a",
		SessionID: "session-a",
	}, "")
	if err == nil {
		t.Fatalf("expected local scope validation error")
	}
}

func TestEvaluateDecisionIsRetiredLocally(t *testing.T) {
	client := New("http://127.0.0.1:1")
	_, err := client.EvaluateDecision(SurfaceRecord{"principal": "spoofed"})
	if err == nil || err.Error() != "EvaluateDecision is retired; use EvaluateDecisionWithScope with EvaluationScope" {
		t.Fatalf("unexpected legacy evaluator error: %v", err)
	}
}

func TestEvaluateDecisionWithScopeDoesNotInheritClientWorkspace(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("X-Helm-Workspace-ID"); got != "" {
			t.Fatalf("unexpected inherited workspace header: %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("X-Helm-Receipt-ID", "rcpt-decision-2")
		_, _ = w.Write([]byte(`{"id":"decision-2"}`))
	}))
	defer server.Close()

	client := New(server.URL, WithAPIKey("token"), WithWorkspaceID("workspace-default"))
	_, err := client.EvaluateDecisionWithScope(
		DecisionRequest{Action: "read", Resource: "ticket:2"},
		EvaluationScope{TenantID: "tenant-a", PrincipalID: "operator-a", SessionID: "session-a"},
		"",
	)
	if err != nil {
		t.Fatal(err)
	}
}

func TestEvaluateDecisionWithScopeRejectsMissingReceiptID(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"id":"decision-3"}`))
	}))
	defer server.Close()

	client := New(server.URL, WithAPIKey("token"))
	_, err := client.EvaluateDecisionWithScope(
		DecisionRequest{Action: "read", Resource: "ticket:3"},
		EvaluationScope{TenantID: "tenant-a", PrincipalID: "operator-a", SessionID: "session-a"},
		"",
	)
	if err == nil || err.Error() != "evaluator response missing required X-Helm-Receipt-ID" {
		t.Fatalf("unexpected missing receipt error: %v", err)
	}
}

File: sdk/ts/src/evaluate-contract.test.ts
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { HelmClient } from "./client.js";
import type { DecisionRequest } from "./types.gen.js";

function jsonResponse(
  body: unknown,
  headers: Record<string, string> = {},
): Response {
  return new Response(JSON.stringify(body), {
    headers: { "Content-Type": "application/json", ...headers },
  });
}

describe("evaluateDecisionWithScope", () => {
  let fetchSpy: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    fetchSpy = vi.fn();
    vi.stubGlobal("fetch", fetchSpy);
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("binds authenticated scope and sends only the canonical request body", async () => {
    fetchSpy.mockResolvedValue(
      jsonResponse(
        { id: "decision-1", tenant_id: "tenant-a", session_id: "session-a" },
        {
          "X-Helm-Receipt-ID": "rcpt-decision-1",
          "X-Helm-Idempotency-Replayed": "true",
        },
      ),
    );
    const client = new HelmClient({
      baseUrl: "http://helm.test",
      apiKey: "token",
      tenantId: "tenant-default",
      principalId: "principal-default",
      workspaceId: "workspace-default",
    });
    const request = {
      action: "read_ticket",
      resource: "ticket:1",
      context: { priority: "low" },
      session_history: [{
        action: "read_history",
        resource: "ticket:0",
        verdict: "ALLOW",
        timestamp: 1,
        principal: "spoofed-history-principal",
      }],
      principal: "spoofed-principal",
    } as DecisionRequest;

    const result = await client.evaluateDecisionWithScope(
      request,
      {
        tenantId: "tenant-a",
        principalId: "principal-a",
        sessionId: "session-a",
        workspaceId: "workspace-a",
      },
      "request-1",
    );

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toBe("http://helm.test/api/v1/evaluate");
    expect(init.headers).toMatchObject({
      Authorization: "Bearer token",
      "X-Helm-Tenant-ID": "tenant-a",
      "X-Helm-Principal-ID": "principal-a",
      "X-Helm-Session-ID": "session-a",
      "X-Helm-Workspace-ID": "workspace-a",
      "Idempotency-Key": "request-1",
    });
    expect(JSON.parse(init.body as string)).toEqual({
      action: "read_ticket",
      resource: "ticket:1",
      context: { priority: "low" },
      session_history: [{
        action: "read_history",
        resource: "ticket:0",
        verdict: "ALLOW",
        timestamp: 1,
      }],
    });
    expect(result.decision.id).toBe("decision-1");
    expect(result.receiptId).toBe("rcpt-decision-1");
    expect(result.replayed).toBe(true);
  });

  it("rejects missing authentication or scope before making a request", async () => {
    const request: DecisionRequest = {
      action: "read_ticket",
      resource: "ticket:1",
    };
    const clientWithoutKey = new HelmClient({ baseUrl: "http://helm.test" });
    await expect(
      clientWithoutKey.evaluateDecisionWithScope(request, {
        tenantId: "tenant-a",
        principalId: "principal-a",
        sessionId: "session-a",
      }),
    ).rejects.toThrow("apiKey is required");

    const client = new HelmClient({
      baseUrl: "http://helm.test",
      apiKey: "token",
    });
    await expect(
      client.evaluateDecisionWithScope(request, {
        tenantId: "tenant-a",
        principalId: "principal-a",
        sessionId: "",
      }),
    ).rejects.toThrow("sessionId is required");
    expect(fetchSpy).not.toHaveBeenCalled();
  });

  it("does not send a configured workspace when the explicit scope omits it", async () => {
    fetchSpy.mockResolvedValue(
      jsonResponse(
        { id: "decision-2", tenant_id: "tenant-a", session_id: "session-a" },
        { "X-Helm-Receipt-ID": "rcpt-decision-2" },
      ),
    );
    const client = new HelmClient({
      baseUrl: "http://helm.test",
      apiKey: "token",
      workspaceId: "workspace-default",
    });

    await client.evaluateDecisionWithScope(
      { action: "read_ticket", resource: "ticket:2" },
      { tenantId: "tenant-a", principalId: "principal-a", sessionId: "session-a" },
    );

    const [, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(init.headers).not.toHaveProperty("X-Helm-Workspace-ID");
  });

  it("rejects a successful evaluator response without its required receipt ID", async () => {
    fetchSpy.mockResolvedValue(jsonResponse({ id: "decision-3" }));
    const client = new HelmClient({ baseUrl: "http://helm.test", apiKey: "token" });

    await expect(
      client.evaluateDecisionWithScope(
        { action: "read_ticket", resource: "ticket:3" },
        { tenantId: "tenant-a", principalId: "principal-a", sessionId: "session-a" },
      ),
    ).rejects.toThrow("missing required X-Helm-Receipt-ID");
  });

  it("rejects the retired generic evaluator without making a request", async () => {
    const client = new HelmClient({ baseUrl: "http://helm.test" });
    await expect(client.evaluateDecision({ principal: "spoofed" })).rejects.toThrow("evaluateDecision is retired");
    expect(fetchSpy).not.toHaveBeenCalled();
  });
});
